tests/end2end/inprocess/test.py

```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/env python3
+
+import copy
+import os
+
+import torch
+
+import git_theta
+
+print("loading old model")
+model = torch.load("og_model.pt")
+print("making a copy of the model with a single value change")
+updated_model = copy.deepcopy(model)
+updated_model["layers.0.hidden.weight"] = torch.rand(
+    *updated_model["layers.0.hidden.weight"].shape
+)
+
+print("committing the same model to different paths")
+model_1_sha = git_theta.save(model, "model_1.pt")
+model_2_sha = git_theta.save(model, "model_2.pt")
+print("committing the changed model to the same path.")
+model_1_1_sha = git_theta.save(updated_model, "model_1.pt")
+
+print("Making sure the models we not saved to disk.")
+assert not os.path.exists("model_1.pt")
```

Comment:
Re: your comment above: "TODO: Should we have a `materialize` flag that writes the model to disk so you
# don't have git status showing a diff?" It kind of creeps me out that are referring to things via paths that aren't actually ever materialized on disk...

Reply:
They do get materialized on disk, it's just in the "git future" so they do live on disk, but it's a question of if the git working directory has the right commit checked out. So the path would be materialized in the working tree if you checked out the next commit, but it isn't checked out right now.

I think the main question is just should the checkout step be an optional part of the save/load function (if we force it we would do a bunch of extra I/O we don't need to) or if the caller should just make a checkout afterwards.

Reply:
Should we have an arg (true by default seems to make sense to me) that specifies whether the checkout should be done after a save/load?

Reply:
I added flags that will keep the on-disk data sync'd to the in-memory model

Reply:
... should they default to True?

Reply:
I don't think they should default to True, this isn't the way we would implement save/load if the main goal is keeping memory and disk in sync. If we wanted to do that our implementations would be:

```python
def save(...):
  ckpt.save(path)
  repo.git.commit(msg)
  
def load(...):
  repo.git.checkout(sha_or_tag)
  ckpt.load(path)
```

When you don't require the working tree has to be in sync with the in-memory then you only need 1 write to git storage for saving and 1 read from git storage for loading. When you force the checkout syncing you add an extra read/write of the model so you end up with the same cost as the simple implementations above (which are really basically wrappers around making/checking out commits, and seems less ergonomic than using git directly).

My thought was that this api would be used for things like, as you are training you make a commit/tag each time your model does better on the dev set and then at the end you load the best performing model and run it on the test set. Since this is a long running program but automatic program we don't care that the working tree is different. Then once the training is done, you can jump around in git history to get different model checkpoints instead of having multiple copies indexed by the step.

Reply:
I see. I guess I just worry about someone running a .save and expecting that this implies that the checkpoint is actually saved to disk where they expect it to be (as would happen with a torch.save or whatever). Not sure what the right way to communicate that this is *not* the case.

Reply:
I could add a log warning/info that is like 

```
You have saved a model to ${path} directly in git. You are no longer at HEAD, run `git checkout ${sha}` or `git checkout HEAD` to see the model in the working tree.
```

I can see how people might be confused, but just seems weird to have functions that are basically pointless unless non-default fault flags are set lol

Reply:
Yeah, I understand the point that these functions are only beneficial/make sense with "False" default, but I think the "False" behavior is probably not what users are expecting from functions called "save"/"load". Is there any way to do the git-to-working-directory copy asynchronously (i.e. not blocking the Python process that did the save/load )or is that a bad idea? Otherwise, maybe it would be clearer if the functions were called "write_to_git"/"load_from_git" or something? Or maybe I am making wrong assumptions about user expectations, not sure.

Reply:
Good idea, I renamed them to `save_to_git` and `load_from_git`